lib/elixir/src/elixir_expand.erl

```diff
@@ -696,6 +697,20 @@ assert_no_ambiguous_op(Name, Meta, [Arg], E) ->
 assert_no_ambiguous_op(_Atom, _Meta, _Args, _E) ->
   ok.
 
+assert_no_clauses(_Name, _Meta, [], _E) ->
+  ok;
+assert_no_clauses(Name, Meta, Args, E) ->
+  assert_arg_with_no_clauses(Name, Meta, lists:last(Args), E).
+
+assert_arg_with_no_clauses(_Name, _Meta, Arg, _E) when not is_list(Arg) ->
+  ok;
+assert_arg_with_no_clauses(_Name, _Meta, [], _E) ->
+  ok;
+assert_arg_with_no_clauses(Name, Meta, [{Opt, [{'->', _, _} | _]}], E) when Opt =/= rescue, Opt =/= 'catch' ->
```

Comment:
From what I said above, if we find a `->` during expand_local/expand_remote, then it should always fail, regardless of the context.